feat(tiller): add Files map to templates

technosophos opened the pull request:
Templates can now access the non-template files in a chart by using
the '{{.Files}}' map inside of a template.

Relates to #950

technosophos committed:
feat(tiller): add Files map to templates

Templates can now access the non-template files in a chart by using
the '{{.Files}}' map inside of a template.

Relates to #950

technosophos commented:
As @michelleN noted, there is one glaring oversight in this fix: There is no way to access the files by name if they contain a slash or a dot (and other non alphanum chars), as there is no map access syntax in Go templates.

technosophos commented:
Okay, I have corrected the main oversight. It is now easy to access files by name, even if the name contains special characters.

Usage:

1. Add a file to your chart, say "foo/bar.txt"
2. Reference this file in a template: `{{.Files.Get "foo/bar.txt" | printf "%s"}}`
3. Execute the chart

The output of `Files.Get` is a `[]byte` (so that we can support binary files). So to render them as text, the body should be passed through `printf "%s"`.

prydonius commented on pkg/chartutil/values_test.go:
```diff
@@ -106,6 +110,18 @@ where:
+	if name := res["Release"].(map[string]interface{})["Name"]; fmt.Sprint(name) != "Seven Voyages" {
+		t.Errorf("Expected release name 'Seven Voyages', got %q", name)
+	}
+	if data := res["Files"].(Files)["scheherazade/shahryar.txt"]; string(data) != "1,001 Nights" {
```
Should this test the `Get` function directly rather than the map?

technosophos replied:
I guess I should probably split Files into its own file, and then add some specific tests.

prydonius commented on pkg/chartutil/values.go:
```diff
@@ -90,6 +91,31 @@ func (v Values) Encode(w io.Writer) error {
 	return err
 }
 
+// Files is a map of files in a chart that can be accessed from a template.
```
Should this be in it's own `files.go` file?

prydonius commented:
Could files be rendered as partials? One of the issues with the `{{ template }}` function is I can't pipe it to `indent`, where as it sounds like it's possible with this?

technosophos commented:
Moved this stuff to `files.go` and added `files_test.go` with tests specifically agains the struct.

I see what you mean about `template`. I am going to look into that. Maybe a separate issue for that one?

prydonius commented on pkg/chartutil/files.go:
```diff
@@ -0,0 +1,45 @@
+//
+// This is intended to be accessed from within a template, so a missed key returns
+// an empty []byte.
+func (f Files) Get(name string) []byte {
```
Wondering if there should be a `GetString` alternative, as `{{.Files.Get "foo/bar.txt" | printf "%s"}}` can get tedious and long?

technosophos replied:
Good idea. Adding that.

technosophos commented:
Added `Files.GetString(name string)`, updated tests accordingly.

prydonius commented:
sweet!
